dotcom-rendering/src/components/AudioPlayer/AudioPlayer.tsx

```diff
@@ -0,0 +1,355 @@
+import { log } from '@guardian/libs';
+import type { AudioEvent, TAudioEventType } from '@guardian/ophan-tracker-js';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { getOphan } from '../../client/ophan/ophan';
+import { Playback } from './components/Playback';
+import { ProgressBar } from './components/ProgressBar';
+import { CurrentTime, Duration } from './components/time';
+import { Volume } from './components/Volume';
+import { Wrapper } from './components/Wrapper';
+
+// possible events for audio in ophan
+type AudioEvents = TAudioEventType extends `audio:content:${infer E}`
+	? E
+	: never;
+
+// possible progress events for audio in ophan
+type AudioProgressEvents = Extract<
+	AudioEvents,
+	`${number}`
+> extends `${infer N extends number}`
+	? N
+	: never;
+
+type AudioPlayerProps = {
+	/** The audio source you want to play. */
+	src: string;
+	/**
+	 * Optional, pre-computed duration of the audio source.
+	 * If it's not provided it will be calculated once the audio is loaded.
+	 */
+	duration?: number;
+	/**
+	 * Optionally hide the volume controls if setting the volume is better
+	 * handled elsewhere, e.g on a mobile device.
+	 */
+	showVolumeControls?: boolean;
+	/** media element ID for Ophan */
+	mediaId: string;
+};
+
+/**
+ * Audio player component.
+ */
+export const AudioPlayer = ({
+	src,
+	duration: preCalculatedDuration,
+	showVolumeControls = true,
+	mediaId,
+}: AudioPlayerProps) => {
+	// ********************* ophan stuff *********************
+
+	// we'll send listening progress reports to ophan at these percentage points
+	// through playback (100% is handled by the 'ended' event)
+	const audioProgressEvents = useRef<Set<AudioProgressEvents>>(
+		new Set([25, 50, 75]),
+	);
+
+	// wrapper to send audio events to ophan
+	const reportAudioEvent = useCallback(
```

Comment:
It's not a hard rule but it's a convention to have all the useState hooks first. It's possible in future `reportAudioEvent` might need to rely on state so it would need to be moved down in order to avoid temporal dead zone issues.

Reply:
resolved by #12671